### PR TITLE
8298377: JfrVframeStream causes deadlocks in ZGC

### PR DESCRIPTION
**Pull request by fisk**
The JfrVFrameStream is used while generating stack traces for events. One of the events are the ZPage allocation event. This event is sometimes sent when ZGC is relocating. The current implementation of JfrVFrameStream uses WalkContinuation::include, which causes JFR to walk the continuation and perform GC barriers. This is problematic, since ZGC has a requirement that we never perform load barriers while running the relocation code. If we do, we might end up performing other reloctions from the the relocation code, and in some cases that causes dead locks. 

I propose that JFR doesn't walk the stack during stack watermark processing. This PR implements that change. This PR replaces https://github.com/openjdk/jdk/pull/11586

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298377](https://bugs.openjdk.org/browse/JDK-8298377): JfrVframeStream causes deadlocks in ZGC

### Reviewers
 * [Markus Grönlund](https://openjdk.org/census#mgronlun) (@mgronlun - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11778/head:pull/11778` \
`$ git checkout pull/11778`

Update a local copy of the PR: \
`$ git checkout pull/11778` \
`$ git pull https://git.openjdk.org/jdk pull/11778/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11778`

View PR using the GUI difftool: \
`$ git pr show -t 11778`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11778.diff">https://git.openjdk.org/jdk/pull/11778.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/runtime/disableStackTracingMark.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+#include "precompiled.hpp"
+#include "runtime/disableStackTracingMark.hpp"
+#include "runtime/javaThread.hpp"
+
+DEBUG_ONLY(THREAD_LOCAL bool DisableStackTracingMark::_is_active = false;)
+
+DisableStackTracingMark::DisableStackTracingMark(JavaThread* jt)
+  : _jt(nullptr),
+    _sp(nullptr) {
+  if (jt == Thread::current()) {
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/disableStackTracingMark.cpp:34]: Just a drive-by comment, but what is the case for `_jt` not being the current thread?

Comment (fisk) [reply src/hotspot/share/runtime/disableStackTracingMark.cpp:34]: If one thread is processing oops of a different thread, for example when you perform a handshake with a target thread and its oops need to be sane, then the jt is the target thread, which is not the current one. But the stack sampling is based on the current thread always, so if the passed in thread isn't the current thread, there is nothing to do. Hmm - I should add a comment shouldn't I? :-)

**Comment by mgronlun** [comment]
The upshot with this solution compared to not having stacktraces for the ZPage Allocation Event is that only the relocation parts are sensitive? Other allocation sites can still have stacktraces?

File: src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp [hunk 2]
```diff
@@ -387,6 +388,7 @@ bool JfrThreadSampleClosure::do_sample_thread(JavaThread* thread, JfrStackFrame*
   }
 
   bool ret = false;
+  StackWatermarkSet::start_processing(thread, StackWatermarkKind::gc);
```
Comment (mgronlun) [inline src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp:391]: Is this now needed unconditionally, for all threads sampled? What is the overhead of introducing it?

File: src/hotspot/share/runtime/deoptimization.cpp [hunk 3]
```diff
@@ -751,16 +752,16 @@ JRT_LEAF(BasicType, Deoptimization::unpack_frames(JavaThread* thread, int exec_m
...
-  // We set the last_Java frame. But the stack isn't really parsable here. So we
-  // clear it to make sure JFR understands not to try and walk stacks from events
-  // in here.
-  intptr_t* sp = thread->frame_anchor()->last_Java_sp();
-  thread->frame_anchor()->set_last_Java_sp(NULL);
+  {
+    // We set the last_Java frame. But the stack isn't really parsable here. So we
+    // clear it to make sure JFR understands not to try and walk stacks from events
+    // in here.
+    DisableStackTracingMark dstm(thread);
```
Comment (mgronlun) [inline src/hotspot/share/runtime/deoptimization.cpp:759]: You replace the old system in Deopt, with the new formalized method? Ok.

**Comment by fisk** [comment]
> The upshot with this solution compared to not having stacktraces for the ZPage Allocation Event is that only the relocation parts are sensitive? Other allocation sites can still have stacktraces?

It's even more precise than that. Only relocation events... specifically when processing the stack watermark of the current thread, needs to emit stack traces. Any other relocation event can still print stack traces, and other allocation events as well.

File: src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp (hunk 2 above)
Comment (fisk) [reply src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp:391]: Compared to everything else that thread sampling does, I really wouldn't worry about it. In the common case it will just check if processing already started, and then continue.

Comment (mgronlun) [reply src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp:391]: The reason for asking is for upcoming changes that try to minimize everything involved in sampling. Is this only needed for virtual threads?

Comment (fisk) [reply src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp:391]: I see. It is indeed only needed for virtual threads. If you are going to perform very frequent sampling, then it's worth mentioning that the check if the processing already started requires load_acquire on non-TSO platforms. I suppose that would be the dominating cost. But for this load_acquire to start showing up as a significant overhead, I think you would have to do sampling in the order of magnitude of hundreds of nanosecond interval levels. So unless you are getting to such intense frequencies, I think I would still not worry about it. What do you think?

Comment (mgronlun) [reply src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp:391]: Thanks for the information, Erik. I think we can go with your solution, I will take the information about Virtual Threads into account for later, there is still a lot of magic left that needs to be done in the new sampling system in order to incorporate them.

File: src/hotspot/share/runtime/disableStackTracingMark.hpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+#include "memory/allocation.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/macros.hpp"
+
+class JavaThread;
+
+// Use this class to mark a section of code where stack tracing is not safe
+// and should be avoided.
+class DisableStackTracingMark : public StackObj {
```
Comment (mgronlun) [inline src/hotspot/share/runtime/disableStackTracingMark.hpp:36]: One thing that could perhaps be a bit of a gotcha with this is the name selection, "DisableStackTracingMark". One could get the impression that this disables stack traces entirely. But this is not the case for asynchronous operations, like AsyncGetCallTrace and JFR - they fetch the information from the CPU context to produce a stacktrace, if there is no ljf. Not a biggie, but might be confusing. Perhaps "NoLastJavaFrame" or something?

Comment (fisk) [reply src/hotspot/share/runtime/disableStackTracingMark.hpp:36]: Good point. I'll try to think of a better name.

Comment (fisk) [reply src/hotspot/share/runtime/disableStackTracingMark.hpp]: Trying out ClearFrameAnchorMark in the latest commit - what do you think?

Comment (mgronlun) [reply src/hotspot/share/runtime/disableStackTracingMark.hpp]: I am always slightly jarred about the discrepancy between the API, set_last_java_frame() and reset_last_java_frame() and the fact that the struct field is called _frame_anchor. Conceptually, the "last java frame" is better. Therefore, I would suggest ClearLastJavaFrameMark. Sorry.

Comment (fisk) [reply src/hotspot/share/runtime/disableStackTracingMark.hpp]: Okay, let's take that for a spin.

**Review by mgronlun — APPROVED** [review]
Looks good.

**Comment by fisk** [comment]
Closing in favour of @stefank PR.
